api/lib/infrastructure/repositories/knowledge-element-repository.js

```diff
@@ -89,5 +89,22 @@ module.exports = {
       });
   },
 
+  findByCampaignIdForSharedCampaignParticipation(campaignId) {
+    return BookshelfKnowledgeElement
+      .query((qb) => {
+        qb.select('knowledge-elements.*');
+        qb.leftJoin('campaign-participations', 'campaign-participations.userId', 'knowledge-elements.userId');
+        qb.leftJoin('campaigns', 'campaigns.id', 'campaign-participations.campaignId');
+        qb.innerJoin('target-profiles_skills', function() {
+          this.on('target-profiles_skills.targetProfileId', '=', 'campaigns.targetProfileId')
+            .andOn('target-profiles_skills.skillId', '=', 'knowledge-elements.skillId');
+        });
+      })
+      .where({ 'campaign-participations.isShared': true })
+      .where({ 'campaign-participations.campaignId': campaignId })
+      .where({ status: 'validated' })
```

Comment:
Le résultat de cette requête ne correspond pas aux `knowledge-elements` validés au moment du partage dans la campagne indiquée. En effet, on sélectionne les KE :
* des utilisateurs ayant participé à la campagne (via `campaign-participations`) ;
* ayant partagé leurs résulats (`{ 'campaign-participations.isShared': true }`) ;
* correspondant à un acquis du `target-profile` de la campagne ;
* ayant le statut `validated`.

Plusieurs filtres sont manquants :
* les `knowledge-elements` plus anciens que la date de partage (`campaign-participations.sharedAt`) ne doivent pas être pris en compte ;
* les `knowledge-elements` pour lesquels il existe un autre `knowledge-element` plus récent pour le même utilisateur et le même acquis ne doivent pas être pris en compte. Par exemple si sur un acquis X j'ai des KE aux statuts `validated`, `reset` puis `rejected` alors il ne faut pas compter cet acquis dans les validés. De même si j'ai `validated`, `reset` puis `validated` il ne faut le compter qu'une fois.

Je suggère d'aller regarder le code de calcul des résultats collectifs de campagne avoir un exemple d'une façon correcte de récupérer les `knowledge-elements.

Reply:
@jonathanperret Ouais j'ai suggéré une requête alternative à cette affaire en passant par les `assessments` liés aux `campaign-participations`, et en plus elle s'en sort un peu mieux.

Reply:
Les `assessments` ? 😱 
Ça a effectivement des chances d'aller plus vite mais ça n'a pas l'air non plus d'une réponse correcte. En effet il faut tenir compte des `knowledge-elements` acquis par l'utilisateur en-dehors de la campagne ! Sinon quelqu'un qui aurait travaillé toutes les compétences en autonomie avant la campagne (ou pendant, mais avant de partager ses résultats), et à qui on n'aurait donc posé aucune question lors de sa participation à la campagne, serait considéré comme n'ayant aucun acquis validé…

Reply:
@jonathanperret Je suis pas sûr de comprendre les filtres manquants

> les knowledge-elements plus anciens que la date de partage (campaign-participations.sharedAt) ne doivent pas être pris en compte ;

Ce qui a été validé avant le partage des résultats n'est pas pris en compte ? C’est pas l'inverse ? Il faut prendre en compte uniquement les knowledge elements qui ont été validés avant la date de partage non ?

> les knowledge-elements pour lesquels il existe un autre knowledge-element plus récent pour le même utilisateur et le même acquis ne doivent pas être pris en compte. Par exemple si sur un acquis X j'ai des KE aux statuts validated, reset puis rejected alors il ne faut pas compter cet acquis dans les validés. De même si j'ai validated, reset puis validated il ne faut le compter qu'une fois.

Je reformule pour savoir si j'ai compris : Je peux évaluer un acquis X plusieurs fois donc je peux "l'avoir" puis le "perdre" d'où la possibilité d'avoir plusieurs knowledge elements pour un utilisateur sur un même acquis avec différents statuts ? 
Donc pour un acquis X je dois prendre les knowledge elements créés avant le partage des résultats de l'utilisateur et dans ces knowledge elements je prend le plus récent ?

Reply:
> @jonathanperret Je suis pas sûr de comprendre les filtres manquants
> 
> > les knowledge-elements plus anciens que la date de partage (campaign-participations.sharedAt) ne doivent pas être pris en compte ;
> 
> Ce qui a été validé avant le partage des résultats n'est pas pris en compte ? C’est pas l'inverse ? Il faut prendre en compte uniquement les knowledge elements qui ont été validés avant la date de partage non ?

_My bad_, je voulais écrire "les plus **récents** ne doivent pas être pris en compte" bien sûr.

> > les knowledge-elements pour lesquels il existe un autre knowledge-element plus récent pour le même utilisateur et le même acquis ne doivent pas être pris en compte. Par exemple si sur un acquis X j'ai des KE aux statuts validated, reset puis rejected alors il ne faut pas compter cet acquis dans les validés. De même si j'ai validated, reset puis validated il ne faut le compter qu'une fois.
> 
> Je reformule pour savoir si j'ai compris : Je peux évaluer un acquis X plusieurs fois donc je peux "l'avoir" puis le "perdre" d'où la possibilité d'avoir plusieurs knowledge elements pour un utilisateur sur un même acquis avec différents statuts ?
> Donc pour un acquis X je dois prendre les knowledge elements créés avant le partage des résultats de l'utilisateur et dans ces knowledge elements je prend le plus récent ?

Exactement. C'est ce qui est fait systématiquement quand on veut connaître le profil d'un utilisateur à une date donnée.

Reply:
@jonathanperret  D'acc merci beaucoup pour les explications et les conseils !

Reply:
@jonathanperret Raaaaah c'est vrai bon sang !! Purée, j'oublie à chaque fois !

Reply:
@laura-bergoens Faut reconnaître que c'est pas simple aussi. J'étais en train de me demander ce qu'on pourrait faire pour avoir moins de chances de se tromper.

Une idée qui m'est venue c'est que le nom `knowledge-elements` n'évoque pas du tout le fait qu'ils sont immuables, datés et remplaçables.

Peut-être qu'un nouveau nom aiderait à tomber plus souvent dans le [Pit of success](https://medium.com/@ricomariani/the-pit-of-success-cfefc6cb64c8). Par exemple, `user-skill-events`, qui évoquerait le fait qu'il s'agit d'une association entre un utilisateur et un acquis, et que ce sont des événements datés immuables, pas un statut exploitable tel quel. Qu'en dis-tu ? 🤔